File: lib/java/opentoken/src/test/java/com/truveta/opentoken/io/csv/PersonAttributesCSVWriterTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.csv;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.nio.file.Files;
import java.util.LinkedHashMap;
import java.util.Map;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PersonAttributesCSVWriterTest {

    private File tempFile;
    private PersonAttributesCSVWriter writer;

    @BeforeEach
    void setUp() throws Exception {
        tempFile = Files.createTempFile("test_person_attributes_", ".csv").toFile();
        writer = new PersonAttributesCSVWriter(tempFile.getAbsolutePath());
    }

    @AfterEach
    void tearDown() throws Exception {
        if (writer != null) {
            writer.close();
        }
        if (tempFile.exists()) {
            tempFile.delete();
        }
    }

    @Test
    void testWriteSingleRecord() throws Exception {
        Map<String, String> data = new LinkedHashMap<>();
        data.put("RecordId", "123");
        data.put("Name", "John Doe");
        data.put("SocialSecurityNumber", "123-45-6789");

        writer.writeAttributes(data);
        writer.close();

        try (BufferedReader reader = new BufferedReader(new FileReader(tempFile))) {
            String header = reader.readLine();
            assertEquals("RecordId,Name,SocialSecurityNumber", header);

            String record = reader.readLine();
            assertEquals("123,John Doe,123-45-6789", record);
        }
    }

    @Test
    void testWriteMultipleRecords() throws Exception {
        Map<String, String> data1 = new LinkedHashMap<>();
        data1.put("RecordId", "123");
        data1.put("Name", "John Doe");
        data1.put("SocialSecurityNumber", "123-45-6789");

        Map<String, String> data2 = new LinkedHashMap<>();
        data2.put("RecordId", "456");
        data2.put("Name", "Jane Smith");
        data2.put("SocialSecurityNumber", "987-65-4321");

        writer.writeAttributes(data1);
        writer.writeAttributes(data2);
        writer.close();

        try (BufferedReader reader = new BufferedReader(new FileReader(tempFile))) {
            String header = reader.readLine();
            assertEquals("RecordId,Name,SocialSecurityNumber", header);

            String record1 = reader.readLine();
            assertEquals("123,John Doe,123-45-6789", record1);

            String record2 = reader.readLine();
            assertEquals("456,Jane Smith,987-65-4321", record2);
        }
    }

    @Test
    void testHeaderWrittenOnlyOnce() throws Exception {
        Map<String, String> data1 = new LinkedHashMap<>();
        data1.put("RecordId", "123");
        data1.put("Name", "John Doe");

        Map<String, String> data2 = new LinkedHashMap<>();
        data2.put("RecordId", "456");
        data2.put("Name", "Jane Smith");

        writer.writeAttributes(data1);
        writer.writeAttributes(data2);
        writer.close();

        try (BufferedReader reader = new BufferedReader(new FileReader(tempFile))) {
            String header = reader.readLine();
            assertEquals("RecordId,Name", header);

            String record1 = reader.readLine();
            assertEquals("123,John Doe", record1);

            String record2 = reader.readLine();
            assertEquals("456,Jane Smith", record2);
        }
    }
}

File: lib/python/opentoken/src/test/opentoken/io/csv/person_attributes_csv_writer_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import os
import tempfile

import pytest

from opentoken.io.csv.person_attributes_csv_writer import PersonAttributesCSVWriter


class TestPersonAttributesCSVWriter:
    """Test cases for PersonAttributesCSVWriter."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.temp_file = tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False)
        self.temp_file_path = self.temp_file.name
        self.temp_file.close()
        self.writer = PersonAttributesCSVWriter(self.temp_file_path)

    def teardown_method(self):
        """Clean up after each test method."""
        if self.writer:
            self.writer.close()
        if os.path.exists(self.temp_file_path):
            os.unlink(self.temp_file_path)

    def test_write_single_record(self):
        """Test writing a single record to CSV."""
        data = {
            "RecordId": "123",
            "Name": "John Doe",
            "SocialSecurityNumber": "123-45-6789"
        }

        self.writer.write_attributes(data)
        self.writer.close()

        with open(self.temp_file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            header = lines[0].strip()
            assert header == "RecordId,Name,SocialSecurityNumber"

            record = lines[1].strip()
            assert record == "123,John Doe,123-45-6789"

    def test_write_multiple_records(self):
        """Test writing multiple records to CSV."""
        data1 = {
            "RecordId": "123",
            "Name": "John Doe",
            "SocialSecurityNumber": "123-45-6789"
        }

        data2 = {
            "RecordId": "456",
            "Name": "Jane Smith",
            "SocialSecurityNumber": "987-65-4321"
        }

        self.writer.write_attributes(data1)
        self.writer.write_attributes(data2)
        self.writer.close()

        with open(self.temp_file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            header = lines[0].strip()
            assert header == "RecordId,Name,SocialSecurityNumber"

            record1 = lines[1].strip()
            assert record1 == "123,John Doe,123-45-6789"

            record2 = lines[2].strip()
            assert record2 == "456,Jane Smith,987-65-4321"

    def test_header_written_only_once(self):
        """Test that header is written only once."""
        data1 = {
            "RecordId": "123",
            "Name": "John Doe"
        }

        data2 = {
            "RecordId": "456",
            "Name": "Jane Smith"
        }

        self.writer.write_attributes(data1)
        self.writer.write_attributes(data2)
        self.writer.close()

        with open(self.temp_file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            header = lines[0].strip()
            assert header == "RecordId,Name"

            record1 = lines[1].strip()
            assert record1 == "123,John Doe"

            record2 = lines[2].strip()
            assert record2 == "456,Jane Smith"